pkg/jobs/registry_test.go

```diff
@@ -973,3 +973,133 @@ func TestRunWithoutLoop(t *testing.T) {
+				},
+			},
+		},
+	})
+	defer s.Stopper().Stop(ctx)
+
+	r := s.JobRegistry().(*Registry)
+
+	resumeStartChan := make(chan struct{})
+	resumeErrChan := make(chan error)
+	defer close(resumeErrChan)
+	RegisterConstructor(jobspb.TypeImport, func(_ *Job, cs *cluster.Settings) Resumer {
+		return FakeResumer{
+			OnResume: func(ctx context.Context) error {
+				resumeStartChan <- struct{}{}
+				return <-resumeErrChan
+			},
+		}
+	})
+
+	currentlyIdle := r.MetricsStruct().JobMetrics[jobspb.TypeImport].CurrentlyIdle
+
+	createJob := func() *Job {
+		jobID := r.MakeJobID()
+		require.NoError(t, kvDB.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+			_, err := r.CreateJobWithTxn(ctx, Record{
+				Details:  jobspb.ImportDetails{},
+				Progress: jobspb.ImportProgress{},
+			}, jobID, txn)
+			return err
+		}))
+		job, err := r.LoadJob(ctx, jobID)
+		require.NoError(t, err)
+		<-resumeStartChan
+		return job
+	}
+
+	tdb := sqlutils.MakeSQLRunner(sqlDB)
+	waitUntilStatus := func(t *testing.T, jobID jobspb.JobID, status Status) {
+		tdb.CheckQueryResultsRetry(t,
+			fmt.Sprintf("SELECT status FROM [SHOW JOBS] WHERE job_id = %d", jobID),
+			[][]string{{string(status)}})
+	}
+
+	t.Run("MarkIdle", func(t *testing.T) {
+		job1 := createJob()
+		job2 := createJob()
+
+		require.False(t, r.IsJobIdle(job1.ID()))
+
+		r.MarkIdle(job1, true)
+		r.MarkIdle(job2, true)
+		require.True(t, r.IsJobIdle(job1.ID()))
+		require.Equal(t, int64(2), currentlyIdle.Value())
+
+		// Repeated calls should not increase metric
+		r.MarkIdle(job1, true)
+		r.MarkIdle(job1, true)
+		require.Equal(t, int64(2), currentlyIdle.Value())
+
+		r.MarkIdle(job1, false)
+		require.Equal(t, int64(1), currentlyIdle.Value())
+		require.False(t, r.IsJobIdle(job1.ID()))
+		r.MarkIdle(job2, false)
+		require.Equal(t, int64(0), currentlyIdle.Value())
+
+		// Let the jobs complete
+		resumeErrChan <- nil
+		resumeErrChan <- nil
+	})
+
+	t.Run("idleness disabled on state updates", func(t *testing.T) {
+		for _, test := range []struct {
+			name   string
+			update func(jobID jobspb.JobID)
+		}{
+			{"pause", func(jobID jobspb.JobID) {
+				resumeErrChan <- MarkPauseRequestError(errors.Errorf("pause error"))
+				waitUntilStatus(t, jobID, StatusPaused)
+			}},
+			{"succeeded", func(jobID jobspb.JobID) {
+				resumeErrChan <- nil
+				waitUntilStatus(t, jobID, StatusSucceeded)
+			}},
+			{"failed", func(jobID jobspb.JobID) {
+				resumeErrChan <- errors.Errorf("error")
+				waitUntilStatus(t, jobID, StatusFailed)
+			}},
+			{"cancel", func(jobID jobspb.JobID) {
+				resumeErrChan <- errJobCanceled
+				waitUntilStatus(t, jobID, StatusCanceled)
+			}},
+		} {
+			t.Run(test.name, func(t *testing.T) {
+				job := createJob()
+
+				require.Equal(t, int64(0), currentlyIdle.Value())
+				r.MarkIdle(job, true)
+				require.Equal(t, int64(1), currentlyIdle.Value())
+
+				test.update(job.ID())
+
+				testutils.SucceedsSoon(t, func() error {
+					if r.IsJobIdle(job.ID()) {
```

Comment:
if you cancel the job, isn't it possible that the registry will drop that job from it's in-memory structure?

Reply:
Yep that is possible, but right now I have `IsJobIdle(...)` simply returning `false` if the job isn't currently adopted in the registry (this is mentioned in the comment above the function).  I could also make `IsJobIdle` return a bool + error for if the job ID isn't found.

Reply:
Hm... Perhaps this is a test only function? If it were renamed to TestingIsJobIdle, I'd be more okay w/ that.

Reply:
I'll make it `TestingIsJobIdle` for now and then see if it ends up being necessary for the changefeeds code in that PR.  I was initially using it when manually deciding to place down a protected timestamp record but that'l no longer be necessary.